### PR TITLE
New backups: Changing the metadata format according to the docs.

### PR DESCRIPTION
**Pull request by makingthematrix**
Some more fields were added and their names changed. Please see the docs for details:
https://github.com/wearezeta/documentation/blob/0180afa7073909aef1dd199cd48fcddf4d915038/topics/backup/README.md#metadata-file

#### APK
[Download build #2706](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2706/artifact/build/artifact/wire-dev-PR3012-2706.apk)
[Download build #2709](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2709/artifact/build/artifact/wire-dev-PR3012-2709.apk)
[Download build #2711](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2711/artifact/build/artifact/wire-dev-PR3012-2711.apk)
[Download build #2713](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2713/artifact/build/artifact/wire-dev-PR3012-2713.apk)
[Download build #2717](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2717/artifact/build/artifact/wire-dev-PR3012-2717.apk)
[Download build #2723](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2723/artifact/build/artifact/wire-dev-PR3012-2723.apk)

### CODE REVIEW

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt [hunk 1]
```diff
@@ -69,4 +75,4 @@ class CreateBackUpUseCase(
     }
 }
 
-data class CreateBackUpUseCaseParams(val userId: UserId, val userHandle: String, val password: String)
+data class CreateBackUpUseCaseParams(val userId: UserId, val clientId: ClientId, val userHandle: String, val password: String)
```
Comment (mejdoo) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:78]: avoid using Scala models in the Kotlin code. Change this to: 

`data class CreateBackUpUseCaseParams(val userId: String, val clientId: String, val userHandle: String, val password: String)`

and pass the ids in **String** format instead.

**Review by mejdoo — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/com/waz/zclient/feature/backup/metadata/BackupMetaData.kt [hunk 2]
```diff
@@ -1,11 +1,26 @@
...
+import org.threeten.bp.Instant
+import java.text.SimpleDateFormat
+import java.util.Locale
 
 @Serializable
 data class BackupMetaData(
-    val userId: String = String.empty(),
+    val platform: String = "Android",
+    @SerialName("user_id") val userId: String = String.empty(),
+    val version: String = Config.versionName(),
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/metadata/BackupMetaData.kt:15]: what's the difference between `version` and `backUpVerison`?

File: app/src/main/kotlin/com/waz/zclient/feature/backup/metadata/BackupMetaData.kt [hunk 3]
```diff
@@ -1,11 +1,26 @@
...
+    @SerialName("creation_time") val creationTime: String = timeNow(),
+    @SerialName("client_id") val clientId: String = String.empty(),
+    // Android specific data
     val userHandle: String = String.empty(),
     val backUpVersion: Int = 0
-)
+) {
+    companion object BackupMetaData {
+        private fun timeNow(): String =
+            SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.getDefault()).format(Instant.now().toEpochMilli())
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/metadata/BackupMetaData.kt:24]: this should be extracted to a some sort of date util class

File: app/src/test/kotlin/com/waz/zclient/feature/backup/metadata/MetaDataHandlerTest.kt [hunk 4]
```diff
@@ -22,22 +22,39 @@ class MetaDataHandlerTest : UnitTest() {
...
     private lateinit var jsonConverter: JsonConverter<BackupMetaData>
 
-    private val backUpVersion = 0
+    private val platform = "Android"
     private val userId = UserId.apply(UUID.randomUUID().toString())
+    private val clientId = ClientId.apply(UUID.randomUUID().toString())
+    private val version = "3.54"
+    private val creationTime = "2020-09-08T10:00:00.000Z"
     private val userHandle = "user"
+    private val backUpVersion = 0
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/feature/backup/metadata/MetaDataHandlerTest.kt:31]: primitives can be constants

File: app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt [hunk 5]
```diff
@@ -37,8 +38,22 @@ class CreateBackUpUseCaseTest : UnitTest() {
     private val userHandle = "user"
     private val password = "password"
     private val backUpVersion = 0
-
-    private val metaData = BackupMetaData(userId.str(), userHandle, backUpVersion)
+    private val platform = "Android"
+    private val clientId = ClientId.apply(UUID.randomUUID().toString())
+    private val version = "3.54"
+    private val creationTime = "2020-09-08T10:00:00.000Z"
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt:44]: primitives can be constants

**Review by gizemb — CHANGES_REQUESTED** [review]

File: app/src/main/kotlin/com/waz/zclient/feature/backup/metadata/BackupMetaData.kt (hunk 2 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/metadata/BackupMetaData.kt:15]: `version` is our app version: e.g. "3.54". `backUpVersion` is the version of the backup schema, currently `1`.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/metadata/BackupMetaData.kt (hunk 3 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/metadata/BackupMetaData.kt:24]: I'd like to later refactor it with that utility method from another PR, so for now I don't want spend time on this.

**Review by mejdoo — APPROVED** [review]

**Review by gizemb — APPROVED** [review]

**Comment by makingthematrix** [comment]
build again

**Pull request merged**